### DISCUSSION TITLE
Will the onedrive client work with MFA/Okta?

### DISCUSSION THREAD

**Discussion by rmoorewrs** [abraunegg/onedrive#1306] [Q&A] [answered]
I've been happily using the onedrive client on my Linux machines for almost a year, and now my company has added multi-factor authentication on our Microsoft accounts. I have the option of Google Authenticator, Okta or SMS. 

I just changed my password and went to do the --logout and re-authorize process, but now it fails. So am I just out of luck now? Or am I missing something? 

Would supporting MFA require an enhancement to onedrive in order to work now? Error messages below.

Thanks,

Rich

--------------------------------------------------------------------------------------------------------------------------------
```
ERROR: Microsoft OneDrive API returned an error with the following message:
  Error Message:    HTTP request returned status code 400 (Bad Request)
  Error Reason:     AADSTS54005: OAuth2 Authorization code was already redeemed, please retry with a new valid code or use an existing refresh token.
Trace ID: xxxxxxx-xxxxx-xxxx-xxxx-xxxxxxxxxxxxxxx
Correlation ID: xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx
Timestamp: 2021-02-28 20:23:15Z
  Calling Function: acquireToken()


ERROR: Microsoft OneDrive API returned an error with the following message:
  Error Message:    HTTP request returned status code 400 (Bad Request)
  Error Reason:     AADSTS900144: The request body must contain the following parameter: 'refresh_token'.
Trace ID: xxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
Correlation ID: xxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx
Timestamp: 2021-02-28 20:23:15Z
  Calling Function: acquireToken()


ERROR: Microsoft OneDrive API returned an error with the following message:
  Error Message:    HTTP request returned status code 401 (Unauthorized)
  Error Reason:     Access token is empty.
  Error Timestamp:  2021-02-28T20:23:15
  API Request ID:   xxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx
  Calling Function: init()


ERROR: Check your configuration as your refresh_token may be empty or invalid. You may need to issue a --logout and re-authorise this client.
```

**Comment by abraunegg** [comment, marked as answer]
@rmoorewrs 
Unfortunately the OneDrive API does not support MFA/2FA - thus, the responses from OneDrive are unable to be used.

You are out of luck here - sorry.

**Reply by rmoorewrs** [reply]
OK, well that's a shame. Your client  has worked extremely well, and I'm going to miss using it. Thank you for the quick response.

**Reply by barthiolutheran1** [reply]
I'd like to note, since my org just enabled 2FA with the Microsoft Authenticator app, I did the usual setup on my new installation (23.04), and then started getting messages like this, that sync failed, and that the service couldn't authenticate, etc.

All I had to do was --reauth, which sent me back to the browser, which this time asked for the 2FA code from my phone, and then spat out a URL that has worked without trouble ever since.

**Reply by abraunegg** [reply]
@barthiolutheran1 
Thanks for providing this valuable feedback. I will update the documentation detailing this, as it appears now that Microsoft has fixed their OneDrive API in regards to this.
